ARROW-16243: [C++][Python] Remove Parquet ReadSchemaField method

alexdesiqueira opened the pull request:
As @pitrou started discussing* in [JIRA#ARROW-16243](https://issues.apache.org/jira/browse/ARROW-16243), it seems that `ReadSchemaField()` doesn't do anything different than `ReadColumn()`. This PR removes it and its Python method.
@jorisvandenbossche says that
> (...) both are different, as Parquet counts the final child leaves, while Arrow counts the top-level parent leaves
> The column indices you pass to eg ReadTable are parquet-based column indices.

Please feel free to close it if (not) necessary.

pitrou commented:
For the record, `ReadSchemaField` was added in https://github.com/apache/parquet-cpp/pull/312.
It was originally called by other functions in that PR, but that was later removed.

Still, we should first deprecate the C++ API, not remove it immediately (you can look for the `ARROW_DEPRECATED` macro which will help you do that). Of course, the unused Cython declaration can be removed.

alexdesiqueira commented:
Thank you for taking a look at it, @pitrou!
> Still, we should first deprecate the C++ API, not remove it immediately

Of course, my bad :sweat_smile: I reverted the last commit, added `ARROW_DEPRECATED` to `ReadSchemaField` and removed `read_schema_field`. Let me know if I'm on the right track.
Thanks again!

pitrou commented on python/pyarrow/_parquet.pyx:
```diff
@@ -1421,13 +1421,6 @@ cdef class ParquetReader(_Weakrefable):
                          .ReadColumn(column_index, &out))
         return pyarrow_wrap_chunked_array(out)
 
-    def read_schema_field(self, int field_index):
```
@jorisvandenbossche Do you think it's fine to remove this (undocumented) method? The ParquetReader class doesn't seem to be documented as a public API...

pitrou replied:
@jorisvandenbossche Ping.

jorisvandenbossche replied:
Sorry, missed this ping. Yes, I think it is fine to just remove this. The ParquetReader is indeed not meant as public API, although it is unfortunate (I see now ..) that we actually import this in the pyarrow.parquet namespace, instead of calling it as `_parquet.ParquetReader` (that probably something else we could fix). As a user you should use `ParquetFile`, the `ParquetReader` is quite inconvenient to construct.
